lib/mutant/context.rb

```diff
@@ -1,19 +1,92 @@
 module Mutant
   # An abstract context where mutations can be applied to.
   class Context
-    include Adamantium::Flat, AbstractType, Concord::Public.new(:source_path)
+    include Adamantium::Flat, Concord::Public.new(:scope, :source_path)
+    extend AST::Sexp
 
-    # Root ast node
-    #
-    # @param [Parser::AST::Node] node
+    NAMESPACE_DELIMITER = '::'.freeze
+
+    # Return root node for mutation
     #
     # @return [Parser::AST::Node]
-    abstract_method :root
+    def root(node)
+      nesting.reverse.reduce(node) do |current, scope|
+        self.class.wrap(scope, current)
+      end
+    end
 
     # Identification string
     #
     # @return [String]
-    abstract_method :identification
+    def identification
+      scope.name
+    end
+
+    # Wrap node into ast node
+    #
+    # @param [Class, Module] scope
+    # @param [Parser::AST::Node] node
+    #
+    # @return [Parser::AST::Class]
+    #   if scope is of kind Class
+    #
+    # @return [Parser::AST::Module]
+    #   if scope is of kind module
+    def self.wrap(scope, node)
+      name = s(:const, nil, scope.name.split(NAMESPACE_DELIMITER).last.to_sym)
+      case scope
+      when Class
+        s(:class, name, nil, node)
+      when Module
+        s(:module, name, node)
+      end
+    end
+
+    # Nesting of scope
+    #
+    # @return [Enumerable<Class,Module>]
+    def nesting
+      const = Object
+      name_nesting.each_with_object([]) do |name, nesting|
```

Comment:
We can. And I acutally thought about this and threw the idea away because it feels counter intuitive to rely on iteration `N - 1` to produce result for `N`. AKA the body mutates an lvar.

But once I wrote it down as per your suggestion, it does not feel so weird so I'm okay with `#map`.